akka-stream-tests/src/test/scala/akka/stream/scaladsl/RetryFlowSpec.scala

```diff
@@ -0,0 +1,572 @@
+            5.seconds,
+            0d,
+            3,
+            failAllValuesFlow
+              .via(Utils.delayCancellation(10.seconds))
+              .viaMat(KillSwitches.single[(Try[Int], Int)])(Keep.right))(alwaysRecoveringFunc))(Keep.both)
+        .toMat(TestSink.probe)(Keep.both)
+        .run()
+
+      sink.request(99)
+      source.sendNext(1)
+      sink.expectNoMessage()
+      killSwitch.abort(FailedElem.failed.get)
+      sink.expectError(FailedElem.failed.get)
+    }
+
+    "tolerate killswitch abort after the RetryFlow while on retry spin" in {
+      val ((source, killSwitch), sink) = TestSource
+        .probe[Int]
+        .map(i => (i, i))
+        .via(stuckForeverRetrying)
+        .via(Utils.delayCancellation(10.seconds))
+        .viaMat(KillSwitches.single[(Try[Int], Int)])(Keep.both)
+        .toMat(TestSink.probe)(Keep.both)
+        .run()
+
+      sink.request(99)
+
+      source.sendNext(1)
+      sink.expectNoMessage()
+
+      killSwitch.abort(FailedElem.failed.get)
+      sink.expectError(FailedElem.failed.get)
+    }
+
+    "finish only after processing all elements in stream" in {
+      val (source, sink) = TestSource
+        .probe[(Int, Int)]
+        .via(RetryFlow.withBackoffAndContext(10.millis, 5.seconds, 0d, 3, failEvenValuesFlow)(incrementFailedValues))
+        .toMat(TestSink.probe)(Keep.both)
+        .run()
+
+      sink.request(99)
+
+      source.sendNext(1 -> 0)
+      source.sendNext(3 -> 0)
+      source.sendNext(2 -> 0)
+      source.sendComplete()
+      sink.expectNext(Success(1) -> 0)
+      sink.expectNext(Success(3) -> 0)
+      sink.expectNext(Success(3) -> 1)
+      sink.expectComplete()
+    }
+  }
+
+  "Coordinator" should {
+
+    class ConstructBench[In, Ctx, Out](retryWith: ((In, Ctx), (Out, Ctx)) => Option[(In, Ctx)]) {
+
+      def setup() = {
+        val bidiFlow: BidiFlow[(In, Ctx), (In, Ctx), (Out, Ctx), (Out, Ctx), NotUsed] =
+          BidiFlow.fromGraph(
+            new RetryFlowCoordinator[In, Ctx, Out, Ctx](
+              minBackoff = 10.millis,
+              maxBackoff = 1.second,
+              0d,
+              3,
+              retryWith))
+
+        val throughDangerousFlow
+            : FlowWithContext[In, Ctx, Out, Ctx, (TestSubscriber.Probe[(In, Ctx)], TestPublisher.Probe[(Out, Ctx)])] =
+          FlowWithContext.fromTuples(
+            Flow.fromSinkAndSourceMat(TestSink.probe[(In, Ctx)], TestSource.probe[(Out, Ctx)])(Keep.both))
+
+        val inOutFlow: Flow[(In, Ctx), (Out, Ctx), (TestSubscriber.Probe[(In, Ctx)], TestPublisher.Probe[(Out, Ctx)])] =
+          bidiFlow.joinMat(throughDangerousFlow)(Keep.right)
+
+        val ((externalIn, (internalOut, internalIn)), externalOut) =
+          TestSource.probe[(In, Ctx)].viaMat(inOutFlow)(Keep.both).toMat(TestSink.probe[(Out, Ctx)])(Keep.both).run()
+
+        (externalIn, externalOut, internalIn, internalOut)
+      }
+      val (externalIn, externalOut, internalIn, internalOut) = setup()
+    }
```

Comment:
I don't think it is worth doing the more tightly coupled internals testing like this, I'd skip this and only test the observable behavior from the public API above but with test probes for all ins and outs.